apps/src/applab/RecordListener.js

```diff
@@ -0,0 +1,230 @@
+RecordListener.EventType = EventType;
+
+/**
+ * Adds a listener which calls the callback when changes are made to the table.
+ * The callback is called once per changed record.
+ * @param {string} tableName Name of the table to listen to.
+ * @param {function(Object, EventType)} callback Callback to call with the record
+ * and the type of event which happened to it. For create and update events,
+ * the entire new record is included. For delete events, only the id is included.
+ * @returns {boolean} true if adding the listener succeeded, or false if
+ * a listener already existed for the specified table.
+ */
+RecordListener.prototype.addListener = function (tableName, callback) {
+  if (this.tableHandlers_[tableName]) {
+    return false;
+  }
+
+  this.tableHandlers_[tableName] = new TableHandler(tableName, callback);
+
+  return true;
+};
+
+/**
+ * Clears the timeouts and resets internal state.
+ */
+RecordListener.prototype.reset = function () {
+  for (var tableName in this.tableHandlers_) {
+    this.tableHandlers_[tableName].reset();
+  }
+  this.tableHandlers_ = {};
+};
+
+//////////////////////////////////////////////////
+// TableHandler
+//////////////////////////////////////////////////
+
+/**
+ * This class encapsulates all the logic and state required for listening to
+ * changes to a single table.
+ * @param {string} tableName
+ * @param {function} callback
+ * @constructor
+ */
+var TableHandler = function (tableName, callback) {
+  /**
+   * @private {string}
+   */
+  this.tableName_ = tableName;
+
+  /**
+   *  @private {function}
+   */
+  this.callback_ = callback;
+
+  /**
+   * @typedef {Object.<number, string>} IdToJsonMap Map from record id
+   * to a JSON-stringified copy of the corresponding record.
+   */
+
+  /**
+   * IdToJsonMap representing the contents of the table.
+   * We use this to efficiently test whether a record's contents has changed.
+   * @private {IdToJsonMap}
+   */
+  this.idToJsonMap_ = {};
+
+  /**
+   * Timeout id of the pending call to window.setTimeout.
+   * @private {number}
+   */
+  this.timeoutId_ = null;
+
+  this.scheduleNextFetch_();
+};
+
+/**
+ * @private
+ */
+TableHandler.prototype.scheduleNextFetch_ = function () {
+  this.timeoutId_ = setTimeout(function recordListener() {
+    this.fetchRecords_();
+  }.bind(this), RECORD_INTERVAL);
+};
+
+/**
+ * @private
+ */
+TableHandler.prototype.fetchRecords_ = function () {
+  var req = new XMLHttpRequest();
+  req.onreadystatechange = this.handleFetchRecords_.bind(this, req);
+  var url = '/v3/shared-tables/' + Applab.channelId + '/' + this.tableName_;
+  req.open('GET', url, true);
+  req.send();
+};
+
+/**
+ * @param {XMLHttpRequest} req
+ * @private
+ */
+TableHandler.prototype.handleFetchRecords_ = function (req) {
+  var done = XMLHttpRequest.DONE || 4;
+  if (req.readyState !== done) {
+    return;
+  }
+
+  if (req.status < 200 || req.status >= 300) {
+    // Ignore errors. In the future we may want to add monitoring or notify
```

Comment:
It would be really good if the client reduced its polling rate in response to 5XX errors (especially 503) so that if the servers are getting overwhelmed we have a way of pushing back and reducing the polling rate.